tests/test_broker_config.cpp

```diff
@@ -77,3 +86,340 @@ TEST_F(TestBrokerConfig, invalid_cid_value_should_fail)
+    R"( { "dns_sd": { ")" + key + R"(": 0 } } )",
+    R"( { "dns_sd": { ")" + key + R"(": false } } )",
+    R"( { "dns_sd": { ")" + key + R"(": true } } )",
+    R"( { "dns_sd": { ")" + key + R"(": {} } } )",
+    R"( { "dns_sd": { ")" + key + R"(": [] } } )",
+    // Empty string is not valid
+    R"( { "dns_sd": { ")" + key + R"(": "" } } )",
+  };
+  // clang-format on
+
+  for (const auto& invalid_input : kInvalidStrings)
+  {
+    std::istringstream test_stream(invalid_input);
+    EXPECT_EQ(config_.ReadFromStream(test_stream), BrokerConfig::ParseResult::kInvalidSetting)
+        << "Input tested: " << invalid_input;
+  }
+}
+
+void TestBrokerConfig::TestDnsSdValidStringValueHelper(
+    const std::string& key, std::function<std::string(const rdmnet::BrokerSettings&)> value_getter)
+{
+  const std::string kTestString = "Broker String Name From Unit Tests";
+  const std::string kValidConfig =
+      R"( { "dns_sd": { ")" + key + R"(": ")" + kTestString + R"(" } } )";
+
+  std::istringstream test_stream(kValidConfig);
+  EXPECT_EQ(config_.ReadFromStream(test_stream), BrokerConfig::ParseResult::kOk);
+  EXPECT_EQ(value_getter(config_.settings), kTestString);
+}
+
+TEST_F(TestBrokerConfig, invalid_dns_sd_service_instance_name_should_fail)
+{
+  TestDnsSdInvalidStringValueHelper("service_instance_name");
+}
+
+TEST_F(TestBrokerConfig, valid_dns_sd_service_instance_name_parsed_correctly)
+{
+  TestDnsSdValidStringValueHelper(
+      "service_instance_name", [](const auto& settings) { return settings.disc_attributes.dns_service_instance_name; });
+}
+
+TEST_F(TestBrokerConfig, invalid_dns_sd_manufacturer_should_fail)
+{
+  TestDnsSdInvalidStringValueHelper("manufacturer");
+}
+
+TEST_F(TestBrokerConfig, valid_dns_sd_manufacturer_parsed_correctly)
+{
+  TestDnsSdValidStringValueHelper("manufacturer",
+                                  [](const auto& settings) { return settings.disc_attributes.dns_manufacturer; });
+}
+
+TEST_F(TestBrokerConfig, invalid_dns_sd_model_should_fail)
+{
+  TestDnsSdInvalidStringValueHelper("model");
+}
+
+TEST_F(TestBrokerConfig, valid_dns_sd_model_parsed_correctly)
+{
+  TestDnsSdValidStringValueHelper("model", [](const auto& settings) { return settings.disc_attributes.dns_model; });
+}
+
+TEST_F(TestBrokerConfig, invalid_dns_sd_scope_should_fail)
+{
+  TestDnsSdInvalidStringValueHelper("scope");
+}
+
+TEST_F(TestBrokerConfig, valid_dns_sd_scope_parsed_correctly)
+{
+  TestDnsSdValidStringValueHelper("scope", [](const auto& settings) { return settings.disc_attributes.scope; });
+}
+
+void TestBrokerConfig::TestInvalidUnsignedIntValueHelper(const std::string& key)
+{
+  // clang-format off
+  const std::vector<std::string> kInvalidIntStrings = {
+    R"( { ")" + key + R"(": -1000 })",
+    R"( { ")" + key + R"(": -30.3 })",
+    R"( { ")" + key + R"(": 20.3 })",
+    // This one might be a bit brittle.... it will fail if sizeof(unsigned long long) == sizeof(unsigned int)...
+    // Is this possible? Better ideas?
```

Comment:
Could this be tested by using a smaller type and calling something deeper (e.g. SetIfInRange), or by replicating the checks in the deeper function with the smaller type?

Reply:
It could.. but I kind of wanted to avoid testing internal implementations... you're not supposed to do that, y'know.

Reply:
True.